Authentication

StankAnt opened the pull request.

StankAnt committed:
feature: configure passport local strategy

StankAnt committed:
feature: user interface

StankAnt committed:
feature: start signin

BobriK502 committed:
feature: start singup

StankAnt committed:
feature: signin

BobriK502 committed:
feature: catch error

BobriK502 committed:
feature: inserting tests

BobriK502 committed:
feature: added custom errors

StankAnt committed:
feature/signin: temporary errors handle

StankAnt committed:
refactor: user controller

BobriK502 committed:
feature: add hash salt & add some ts configs

StankAnt committed:
feature: verify jwt

BobriK502 committed:
feature: added nodemailer

StankAnt committed:
feature: errors handle

StankAnt committed:
feature: verify confirmed accounts

StankAnt committed:
feature: add field type to error

BobriK502 committed:
feature: add custom error add tests

StankAnt committed:
refactor: db models

StankAnt committed:
tests: db models

StankAnt committed:
Merge pull request #2 from CarFinder/CAR-31_signin

Car 31 signin

BobriK502 committed:
feature: add confirm service and test for it

BobriK502 committed:
feature: added timestamp to user model added decode token

BobriK502 committed:
feture: divide tests

StankAnt committed:
remove: unused route

BobriK502 committed:
merge into auth

StankAnt committed:
Merge pull request #3 from CarFinder/CAR-27_Sign-up

Car 27 sign up

StankAnt committed:
feature: not activated account error

BobriK502 committed:
feature: added default values for users model fields

StankAnt committed:
Merge branch 'auth' of https://github.com/CarFinder/CarFinderApi into auth

StankAnt committed:
feature: api route prefix

StankAnt committed:
fix: unactivated error

StankAnt committed:
tests: start write signin api tests

StankAnt committed:
tests: signin request test

StankAnt committed:
tests: finish api tests

StankAnt committed:
fix: move the passport services to passport dir

StankAnt committed:
fix: email sending

BobriK502 committed:
feture: add confirm network test

BobriK502 committed:
fix: error and jwt error message

AntonNarkevich commented on spec/confirm.ts:
```diff
@@ -0,0 +1,127 @@
+      .end((err, res) => {
+        res.should.have.status(200);
+        assert.equal(
+          'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJjb25maXJtZWQiOnRydWUsImVtYWlsIjoicHVwa2luQG1haWwuY29tIiwiaW1hZ2UiOiIiLCJpbnRlcmZhY2VMYW5nIjoiZW4iLCJuYW1lIjoiSXZhbiIsInN1YnNjcmlwdGlvbiI6dHJ1ZX0.gpgESzsa4rQJdHp4RvQ76U-ezVaFTVhyxjpSG8GqiOA',
```
Weird hardcoded tokens? Can we generate the token here? Or at least use a constant.
+ is this going to work in 100% of scenarios? If I'm not mistaken the token contains the date when it was generated + the expiration date.

AntonNarkevich commented on spec/confirm.ts:
```diff
@@ -0,0 +1,127 @@
+          'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJlbWFpbCI6InB1cGtpbkBtYWlsLhNvbSJ9.Si2nol4q-7SeMzCkyvm94s45CzP7kx3jG4y9OLdGcv4'
+      })
+      .end((err, res) => {
+        res.should.have.status(401);
```
Please use 'http-status' module. Numbers shouldn't be hardcoded.

AntonNarkevich commented on spec/confirm.ts:
```diff
@@ -0,0 +1,127 @@
+        res.should.have.status(401);
+        assert.equal(104, res.body.error.code);
+        assert.equal('Token is invalid', res.body.error.enMessage);
+        assert.equal('Токен не валиден', res.body.error.ruMessage);
```
This is not good. The server shouldn't care about the language.
Ideally, it should return just an error code or some kind of a constant (value from the enum)
Than the frontend decides how to handle the error and which message to display.

AntonNarkevich commented on spec/dbModels.ts:
```diff
@@ -0,0 +1,183 @@
+      });
+
+      await user.comparePassword('password', (err: any, isMatch: boolean) => {
+        chai.assert.isTrue(isMatch);
```
I'd presonally call it 'isMatching'. I don't insist here though.

AntonNarkevich commented on spec/signUp.ts:
```diff
@@ -0,0 +1,153 @@
+        .end((err, res) => {
+          assert.equal(error.data.code, res.body.error.code);
+          assert.equal(error.data.enMessage, res.body.error.enMessage);
+          assert.equal(error.data.ruMessage, res.body.error.ruMessage);
```
What if we'd need the Spanish language? Should we refactor all the server code?

AntonNarkevich commented on src/config/config.ts:
```diff
@@ -1 +1,20 @@
 export const db = 'mongodb://Admin:passwordforadmin531@ds163494.mlab.com:63494/car-finder';
```
This should go the the env var and be defaulted to the local value like this:  db = process.env.DB_BLA_BLA || 'mongodb://...';

AntonNarkevich replied:
Same everywhere

BobriK502 replied:
Environment variables added only for production-mode.We are using config from the config object for tests.

AntonNarkevich commented on src/controllers/userController.ts:
```diff
@@ -0,0 +1,45 @@
+  try {
+    await registerUser(ctx.request.body);
+    ctx.status = HttpStatus.CREATED;
+    ctx.body = 'OK';
```
Why do you need to hardcode the text here? 201 is just enough.
Plus this brings in problems with i18n.

AntonNarkevich commented on src/controllers/userController.ts:
```diff
@@ -0,0 +1,45 @@
+
+export const confirmEmail = async (ctx: Koa.Context) => {
+  if (!ctx) {
+    return;
```
Why don't we through here?

AntonNarkevich commented on src/db/schemas/user.ts:
```diff
@@ -1,36 +1,84 @@
+
+UserSchema.post('save', function() {
+  const user = this;
+  sendMail(this.email, this.name);
```
Is this action sync? Doesn't it take a callback?

BobriK502 replied:
As I understand it takes callback if we have sequence of "post" hooks.
I found it here http://mongoosejs.com/docs/middleware.html . As I think we can use post middleware both for async and sync actions.

AntonNarkevich replied:
I mean how the post method knows when the sendMail execution is ended?

AntonNarkevich commented on src/db/schemas/user.ts:
```diff
@@ -1,36 +1,84 @@
+    if (err) {
+      callback(err);
+    }
+    callback(null, isMatch);
```
This is *really bad* the callback is going to be called twice in any case.

AntonNarkevich commented on src/repositories/userRepository.ts:
```diff
@@ -0,0 +1,18 @@
+};
+
+export const get = async (email: string): Promise<IUserModel> => {
+  return (await User.findOne({ email }, { __v: 0 })) as IUserModel;
```
We should get rid of this __v: 0

AntonNarkevich commented on src/services/index.ts:
```diff
@@ -0,0 +1,20 @@
+import * as UserService from './userService';
+
+export const registerUser = async (payload: IUser) => {
+  if (!payload) {
```
Why don't we through here?

AntonNarkevich commented on src/services/userService.ts:
```diff
@@ -0,0 +1,79 @@
+
+export const register = async (payload: IUser) => {
+  if (!payload) {
+    return;
```
Same question. Why don't we thgough?

AntonNarkevich commented on src/services/userService.ts:
```diff
@@ -0,0 +1,79 @@
+};
+
+export const comparePassword = (email: string, password: string, done: any) => {
+  get(email)
```
Why promises? We should use async/await here.

AntonNarkevich commented on src/utils/errors/secureError.ts:
```diff
@@ -0,0 +1,45 @@
+        this.data = {
+          code: codeErrors.INCORRECT_EMAIL_OR_PASS,
+          enMessage: 'Invalid email or password',
+          ruMessage: 'Неверное имя пользователя или пароль',
```
Please stop hardcoding thing like this :)

AntonNarkevich commented on src/utils/index.ts:
```diff
@@ -0,0 +1,50 @@
+    {
+      from: 'Car Fdinder',
+      html,
+      subject: 'Confirma registration',
```
A typo here.

AntonNarkevich commented on src/utils/index.ts:
```diff
@@ -0,0 +1,50 @@
+      to: email
+    },
+    err => {
+      if (err) {
```
Please add logging here. Also this throw won't be handled anywhere and will crash the stack if I'm not mistaken.

AntonNarkevich commented on src/utils/index.ts:
```diff
@@ -0,0 +1,50 @@
+
+const generateEmail = (name: string, email: string, token: string): string => {
+  return `
+  <div style="text-align: center;">
```
We could use any kind of a template engine. E.g. handlebars/pug whatever or something different here.
Although let's keep it like this for now.

BobriK502 replied:
Is it good practice to use template engine for generating a little piece of code?We will not use it anymore.
But if you want we can use pug =) Pug is cool :+1:

AntonNarkevich replied:
Yes, it is a good practice as you might need more email in future.
Personally, I hate pug because of the syntax.

We can keep it as is for now.

BobriK502 committed:
feature : add env ( work only for dev server ) test does not work


fix: pr bugs

bug: add backend validation

fix: databaase error class

fix: signin validation

fix: tests with token and httpStatus and remove usless controller validation


fix: usless validation


tests: validation

refactor: rename isMatch variable

refactor: rename isMatch variable

feature: add env

BobriK502 committed:
fix: async test

AntonNarkevich commented:
Okay, looks much better. Merging in.
